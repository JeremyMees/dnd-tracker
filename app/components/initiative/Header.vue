<script setup lang="ts">
defineEmits<{
  reset: [boolean]
  previous: []
  next: []
}>()

defineProps<{ data: InitiativeSheet | undefined }>()

const resetOpen = ref<boolean>(false)
</script>

<template>
  <div class="relative flex flex-col sm:flex-row gap-x-4 gap-y-2 items-center justify-end container-max w-full">
    <InitiativePet
      v-if="data?.settings?.pet"
      :pet="data.settings.pet"
      class="max-[350px]:hidden absolute top-14 sm:top-6 left-4"
    />
    <div class="flex gap-2 items-center">
      <span class="text-muted-foreground">
        {{ $t('general.round') }}:
        <span class="font-bold text-foreground">
          {{ data?.round || 1 }}
        </span>
      </span>
      <UiPopoverBase v-model:open="resetOpen">
        <UiPopoverTrigger as-child>
          <button
            id="tour-11"
            v-tippy="$t('actions.reset')"
            :disabled="!data?.rows.length"
            aria-label="Reset rounds"
            class="icon-btn-destructive"
          >
            <Icon
              name="tabler:refresh"
              class="text-destructive"
              aria-hidden="true"
            />
          </button>
        </UiPopoverTrigger>
        <UiPopoverContent class="flex flex-col gap-2">
          <button
            :aria-label="$t('components.encounterTable.reset.soft.title')"
            class="flex flex-col gap-2 text-left hover:bg-muted-foreground/10 p-2 rounded-md transition-colors duration-300 ease-in-out"
            @click="$emit('reset', false), resetOpen = false"
          >
            <span class="font-bold">
              {{ $t('components.encounterTable.reset.soft.title') }}
            </span>
            <span class="text-muted-foreground text-sm">
              {{ $t('components.encounterTable.reset.soft.description') }}
            </span>
          </button>
          <UiSeparatorBase />
          <button
            :aria-label="$t('components.encounterTable.reset.hard.title')"
            class="flex flex-col gap-2 text-left hover:bg-muted-foreground/10 p-2 rounded-md transition-colors duration-300 ease-in-out"
            @click="$emit('reset', true), resetOpen = false"
          >
            <span class="font-bold">
              {{ $t('components.encounterTable.reset.hard.title') }}
            </span>
            <span class="text-muted-foreground text-sm">
              {{ $t('components.encounterTable.reset.hard.description') }}
            </span>
          </button>
        </UiPopoverContent>
      </UiPopoverBase>
    </div>
    <div
      id="tour-1"
      class="flex gap-2 items-center bg-primary/50 rounded-lg border-4 border-primary"
    >
      <button
        v-tippy="{ content: $t('actions.prev') }"
        :disabled="data?.round === 1 && data?.activeIndex === 0"
        :aria-label="$t('actions.prev')"
        class="group disabled:cursor-not-allowed duration-300 ease-in-out py-1 pl-1 pr-2 border-r-2 border-primary flex flex-col items-center"
        @click="$emit('previous')"
      >
        <Icon
          name="tabler:chevron-left"
          class="group-disabled:opacity-50 size-6"
          aria-hidden="true"
        />
      </button>
      <p class="font-medium px-2">
        {{ $t('components.inputs.initiativeLabel') }}
      </p>
      <button
        v-tippy="{ content: $t('actions.next') }"
        :disabled="!data?.rows.length"
        :aria-label="$t('actions.next')"
        class="group disabled:cursor-not-allowed py-1 pl-2 pr-1 border-l-2 border-primary flex flex-col items-center"
        @click="$emit('next')"
      >
        <Icon
          name="tabler:chevron-right"
          class="group-disabled:opacity-50 size-6"
          aria-hidden="true"
        />
      </button>
    </div>
  </div>
</template>
