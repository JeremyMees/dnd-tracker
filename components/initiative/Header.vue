<script setup lang="ts">
defineEmits<{
  reset: []
  previous: []
  next: []
}>()

defineProps<{ data: InitiativeSheet }>()
</script>

<template>
  <div class="flex flex-col sm:flex-row gap-4 items-center justify-between container-max w-full">
    <div class="flex gap-2 items-center">
      <span class="text-muted-foreground">
        {{ $t('general.round') }}:
        <span class="font-bold text-foreground">
          {{ data.round }}
        </span>
      </span>
      <button
        v-tippy="{ content: $t('actions.reset') }"
        :disabled="!data.rows.length"
        aria-label="Reset rounds"
        class="disabled:opacity-40 disabled:cursor-not-allowed flex flex-col items-center"
        @click="$emit('reset')"
      >
        <Icon
          name="tabler:refresh"
          class="size-6 text-destructive"
          aria-hidden="true"
        />
      </button>
    </div>
    <div
      id="tour-1"
      class="flex gap-2 items-center bg-primary/50 rounded-lg border-4 border-primary"
    >
      <button
        v-tippy="{ content: $t('actions.prev') }"
        :disabled="data.round === 1 && data.activeIndex === 0"
        :aria-label="$t('actions.prev')"
        class="group disabled:cursor-not-allowed duration-300 ease-in-out py-1 pl-1 pr-2 border-r-2 border-primary flex flex-col items-center"
        @click="$emit('previous')"
      >
        <Icon
          name="tabler:chevron-left"
          class="group-disabled:opacity-40 size-6"
          aria-hidden="true"
        />
      </button>
      <p class="uppercase font-bold px-2">
        {{ $t('components.inputs.initiativeLabel') }}
      </p>
      <button
        v-tippy="{ content: $t('actions.next') }"
        :disabled="!data.rows.length"
        :aria-label="$t('actions.next')"
        class="py-1 pl-2 pr-1 border-l-2 border-primary flex flex-col items-center"
        @click="$emit('next')"
      >
        <Icon
          name="tabler:chevron-right"
          class="size-6"
          aria-hidden="true"
        />
      </button>
    </div>
  </div>
</template>
