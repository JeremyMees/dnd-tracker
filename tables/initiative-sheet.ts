import { createColumnHelper, type InitialTableState, type Row } from '@tanstack/vue-table'
import { actionsTable, expandButton, iconButton } from './generate-functions'

import {
  InitiativeTableRowName,
  InitiativeTableRowHealth,
  InitiativeTableRowActions,
  InitiativeTableRowConditions,
  InitiativeTableRowNotes,
  InitiativeTableRowInit,
  InitiativeTableRowAc,
  InitiativeTableRowDeathSaves,
  InitiativeTableRowConcentration,
  InitiativeTableRowModify,
} from '#components'

const columnHelper = createColumnHelper<InitiativeSheetRow>()

interface ColumnOptions {
  sheet: ComputedRef<InitiativeSheet | undefined>
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable>) => Promise<void>
  openQuickInitModal: () => void
}

export function generateColumns({ sheet, update, openQuickInitModal }: ColumnOptions) {
  const { t } = useI18n()

  return [
    columnHelper.display({
      enableGlobalFilter: false,
      id: 'index',
      cell: ({ row }) => h('div', { class: 'flex items-center gap-x-2 max-w-20' }, [
        h('span', { class: 'text-muted-foreground ml-1' }, row.index + 1),
        expandButton({
          content: t(`actions.${row.getIsExpanded() ? 'hide' : 'show'}`),
          expanded: row.getIsExpanded(),
          cb: () => row.toggleExpanded(),
        }),
      ]),
    }),
    columnHelper.accessor('name', {
      header: t('components.encounterTable.headers.name'),
      cell: ({ row }) => h(InitiativeTableRowName, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('initiative', {
      header: () => h('button', { class: 'flex items-center gap-x-2' }, [
        t('components.encounterTable.headers.init'),
        iconButton({
          icon: 'tabler:bolt-filled',
          content: t('components.encounterTable.quick'),
          color: 'warning',
          cb: () => openQuickInitModal(),
          disabled: !sheet.value?.rows.length,
          iconColor: 'text-warning',
        }),
      ]),
      cell: ({ row }) => h(InitiativeTableRowInit, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('health', {
      header: t('components.encounterTable.headers.health'),
      cell: ({ row }) => h(InitiativeTableRowHealth, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('ac', {
      header: t('components.encounterTable.headers.ac'),
      cell: ({ row }) => h(InitiativeTableRowAc, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      id: 'manage',
      header: t('components.encounterTable.headers.manage'),
      cell: ({ row }) => h(InitiativeTableRowActions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('conditions', {
      header: t('components.encounterTable.headers.conditions'),
      cell: ({ row }) => h(InitiativeTableRowConditions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('note', {
      header: t('components.encounterTable.headers.note'),
      cell: ({ row }) => h(InitiativeTableRowNotes, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('deathSaves', {
      header: t('components.encounterTable.headers.deathSaves'),
      cell: ({ row }) => h(InitiativeTableRowDeathSaves, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('concentration', {
      header: t('components.encounterTable.headers.concentration'),
      cell: ({ row }) => h(InitiativeTableRowConcentration, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      enableGlobalFilter: false,
      enableSorting: false,
      id: 'actions',
      cell: ({ row }) => h(InitiativeTableRowModify, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
  ]
}

export function expandedMarkup(row: Row<InitiativeSheetRow>) {
  return actionsTable(row.original)
}

export const initialState: InitialTableState = {}
