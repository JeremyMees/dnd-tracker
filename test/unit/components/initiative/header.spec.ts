import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, vi } from 'vitest'
import Header from '~/components/initiative/Header.vue'
import { sheet } from '~~/test/unit/fixtures/initiative-sheet'

interface Props {
  data: InitiativeSheet | undefined
}

const props: Props = {
  data: sheet,
}

describe('Initiative header', () => {
  it('Should render correctly with required props', async () => {
    const component = await mountSuspended(Header, { props })

    expect(component.html()).toMatchSnapshot()
  })

  it('Should display pet when its active in the settings', async () => {
    const component = await mountSuspended(Header, { props })

    expect(component.find('[data-test-pet]').exists()).toBeTruthy()
  })

  it('Should not display pet when its not active in the settings', async () => {
    const component = await mountSuspended(Header, {
      props: {
        data: {
          ...sheet,
          settings: {
            ...sheet.settings,
            pet: undefined,
          },
        },
      },
    })

    expect(component.find('[data-test-pet]').exists()).toBeFalsy()
  })

  it('Should display correct round number', async () => {
    const component = await mountSuspended(Header, {
      props: {
        data: {
          ...sheet,
          round: 5,
        },
      },
    })

    expect(component.find('[data-test-round]').text()).toBe('5')
  })

  it('Should display default round number when not provided', async () => {
    const component = await mountSuspended(Header, {
      props: {
        data: {
          ...sheet,
          round: 0,
        },
      },
    })

    expect(component.find('[data-test-round]').text()).toBe('1')
  })

  it('Should disable previous button when round is 1 and activeIndex is 0', async () => {
    const component = await mountSuspended(Header, {
      props: {
        data: {
          ...sheet,
          round: 1,
          activeIndex: 0,
        },
      },
    })

    expect(component.find('[data-test-previous]').attributes('disabled')).toBeDefined()
  })

  it('Should disable action buttons when there are no rows', async () => {
    const component = await mountSuspended(Header, {
      props: {
        data: {
          ...sheet,
          rows: [],
        },
      },
    })

    expect(component.find('[data-test-next]').attributes('disabled')).toBeDefined()
    expect(component.find('[data-test-previous]').attributes('disabled')).toBeDefined()
    expect(component.find('[data-test-reset]').attributes('disabled')).toBeDefined()
  })

  it('Should emit next event when next button is clicked', async () => {
    const component = await mountSuspended(Header, { props })

    await component.find('[data-test-next]').trigger('click')

    expect(component.emitted('next')).toBeTruthy()
  })

  it('Should emit previous event when previous button is clicked', async () => {
    const component = await mountSuspended(Header, {
      props: {
        data: {
          ...sheet,
          activeIndex: 2,
        },
      },
    })

    await component.find('[data-test-previous]').trigger('click')

    expect(component.emitted('previous')).toBeDefined()
  })
})
