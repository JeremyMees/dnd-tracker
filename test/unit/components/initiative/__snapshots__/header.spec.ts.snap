// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Initiative header > Should render correctly with required props 1`] = `
"<div class="relative flex flex-col sm:flex-row gap-x-4 gap-y-2 items-center justify-end container-max w-full"><img src="/gifs/cat.gif" loading="lazy" class="size-6 max-[350px]:hidden absolute top-14 sm:top-6 left-4" data-test-pet="">
  <div class="flex gap-2 items-center"><span class="text-muted-foreground">general.round: <span data-test-round="" class="font-bold text-foreground">1</span></span><button id="tour-11" data-test-reset="" aria-label="Reset rounds" class="icon-btn-destructive" type="button" aria-haspopup="dialog" aria-expanded="false" aria-controls="" data-state="closed"><span class="iconify i-tabler:refresh text-destructive" aria-hidden="true"></span></button>
    <!--teleport start-->
    <!--teleport end-->
  </div>
  <div id="tour-1" class="flex gap-2 items-center bg-primary/50 rounded-lg border-4 border-primary"><button data-test-previous="" disabled="" aria-label="actions.prev" class="group disabled:cursor-not-allowed duration-300 ease-in-out py-1 pl-1 pr-2 border-r-2 border-primary flex flex-col items-center"><span class="iconify i-tabler:chevron-left group-disabled:opacity-50 size-6" aria-hidden="true"></span></button>
    <p class="font-medium px-2">components.inputs.initiativeLabel</p><button data-test-next="" aria-label="actions.next" class="group disabled:cursor-not-allowed py-1 pl-2 pr-1 border-l-2 border-primary flex flex-col items-center"><span class="iconify i-tabler:chevron-right group-disabled:opacity-50 size-6" aria-hidden="true"></span></button>
  </div>
</div>"
`;
